components/tidb_query/src/rpn_expr/impl_cast.rs

```diff
@@ -204,14 +288,73 @@ fn cast_any_as_any<From: ConvertTo<To> + Evaluable, To: Evaluable>(
     }
 }
 
-#[rpn_fn(capture = [ctx])]
+#[rpn_fn(capture = [ctx, extra])]
 #[inline]
-pub fn cast_uint_as_int(ctx: &mut EvalContext, val: &Option<Int>) -> Result<Option<i64>> {
+fn cast_uint_as_int(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<i64>> {
     match val {
         None => Ok(None),
         Some(val) => {
-            let val = (*val as u64).to_int(ctx, FieldTypeTp::LongLong)?;
-            Ok(Some(val))
+            if in_union(extra.implicit_args) && *val < 0 {
+                Ok(Some(0))
+            } else {
+                let val = *val as u64;
+                Ok(Some(<u64 as ConvertTo<i64>>::convert(&val, ctx)?))
+            }
+        }
+    }
+}
+
+/// The implementation for push down signature `CastIntAsReal` from unsigned integer.
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_uint_as_real(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<Real>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            if in_union(extra.implicit_args) && *val < 0 {
+                return Ok(Some(Real::new(0f64).unwrap()));
+            }
+            // TODO, TiDB use ConvertIntToUint here, but I think it is a bug
+            let val = *val as u64;
+            let val: u64 = <u64 as ConvertTo<u64>>::convert(&val, ctx)?;
+            Ok(Real::new(val as f64).ok())
+        }
+    }
+}
+
+/// The implementation for push down signature `CastIntAsString` from unsigned integer.
+#[rpn_fn(capture = [ctx, extra])]
+#[inline]
+fn cast_uint_as_string(
+    ctx: &mut EvalContext,
+    extra: &RpnFnCallExtra<'_>,
+    val: &Option<Int>,
+) -> Result<Option<Bytes>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            // TODO, tidb use ConvertIntToUint here, I think it is a bug.
```

Comment:
I think it's a BUG of TiDB (`ConvertIntToUint` will treat the negative number as an overflow but it's a correct representation of uint in int form).